### ISSUE TITLE
Proxy handling mess

### ISSUE THREAD

**Issue by eero-t** [opea-project/GenAIExamples#321]
OPEA services fetch "random" data from internet (HF), and try to access each others' k8s service end points.

Access to Internet from an intranet can be handled by adding `http_proxy`/`https_proxy` environment variables e.g. to "ChatQnA" configMap, but adding `no_proxy` for Kubernetes services does _not_ work so well because Python `urllib`/`urllib3`/`requests` modules used by OPEA service do not handle subnet masks.  I.e. one cannot just specify internal k8s network (`10.0.0.0/8`) for `no_proxy`.

If services use `full` domain name for their k8s service URLs (which is not currently case in the example configMap), cluster domain (`.cluster.local` by default) can be added to`no_proxy`.

However, IMHO better would be to separate OPEA services to:
* ones that pull data from Internet, and
* ones that get data from other k8s services.

As only former need proxy, they could use a separate (or additional) configMap for proxy settings, which would not be used by rest of the services.  That way there's no need to use `no_proxy` for k8s internal acess (+ have documentation for that).

**Comment by eero-t** [comment]
There are a lot of issues about this, https://github.com/opea-project/GenAIExamples/issues/313 is good example of them.

**Comment by eero-t** [comment]
FYI: My current workaround for this mess is specifying k8s service names with full domain name in `configMap`, and adding `.cluster.local` to `no_proxy`.

> IMHO better would be to separate OPEA services  (... and their configMaps)

After that, service end-point env vars could be dropped from configMap, as service name itself is enough.  No IP needed (and if it is, k8s already provides those as env vars, unless that is explicitly disabled).

**Comment by lianhao** [comment]
I don't think using full domain name for their k8s service URLs is appropriate. K8S service name's full FQDN is in the format of <service-name>.<namespace>.svc.<cluster-domain>. Using full FQDN in manifest will prohibit it to be deployed in other namespace, also some operators may change the default cluster-domain(cluster.local) to other values.

I think separate internet access services from other k8s only access services is a good idea, if it's possible.

**Comment by eero-t** [comment]
> I don't think using full domain name for their k8s service URLs is appropriate. K8S service name's full FQDN is in the format of `<service-name>.<namespace>.svc.<cluster-domain>`. Using full FQDN in manifest will prohibit it to be deployed in other namespace, also some operators may change the default cluster-domain(cluster.local) to other values.

Indeed.  Need for such a workaround is why I'm calling the current situation a mess. :-)

**Comment by YuningQiu** [comment]
Hello, many thanks for reaching out and highlighting this matter. We are committed to investigating this issue thoroughly and will provide you with an update as soon as possible. Please be aware that GenAIExamples is currently in active development, and we are dedicated to resolving issues promptly while continuously enhancing our service. Your patience and understanding are greatly appreciated.

**Comment by leslieluyu** [comment]
@eero-t Hi, the manifests files will soon be replaced by a newer version which is generated from helm-chart or you could directly use helm-chart to deploy.
 currently, you could only add `embedding-svc,retriever-svc,reranking-svc,llm-svc` in the no_proxy value of qna_configmap_xeon.yaml.  
that should should work.

**Comment by eero-t** [comment]
> the manifests files will soon be replaced by a newer version which is generated from helm-chart or you could directly use helm-chart to deploy.

@leslieluyu If manifest files are generated from helm-charts **as-is**, project could IMHO drop the manifests and just document how to install Helm, and generate the manifests.  That way they're guaranteed to be in sync...

> currently, you could only add `embedding-svc,retriever-svc,reranking-svc,llm-svc` in the no_proxy value of qna_configmap_xeon.yaml. that should should work.

Right, yes, that should also work.

=> Such `no_proxy` setting could be added to the default configMap, as it should do harm.

**Comment by eero-t** [comment]
> currently, you could only add embedding-svc,retriever-svc,reranking-svc,llm-svc in the no_proxy value of qna_configmap_xeon.yaml.

@leslieluyu These specify the domain the configMap examples:
```
  TEI_EMBEDDING_ENDPOINT: "http://tei-embedding-svc.default.svc.cluster.local:6006"
  TEI_RERANKING_ENDPOINT: "http://tei-reranking-svc.default.svc.cluster.local:8808"
  TGI_LLM_ENDPOINT: "http://tgi-svc.default.svc.cluster.local:9009"
```

In both of these:
* https://github.com/opea-project/GenAIExamples/blob/main/ChatQnA/kubernetes/manifests/qna_configmap_xeon.yaml
* https://github.com/opea-project/GenAIExamples/blob/main/ChatQnA/kubernetes/manifests/qna_configmap_gaudi.yaml

So those should be fixed also in addition to adding the `no_proxy` setting.

**Comment by leslieluyu** [comment]
@eero-t you are right. the "default.svc.cluster.local" part is no use and could be removed.  But even after it was removed. still could not address the problem that a user might want to use a specific namespace. Actually we've already tested how to use script to let users to define the namespace they want. But since we'll soon switch to helm-chart way ...

**Comment by eero-t** [comment]
@leslieluyu You don't need script for setting namespace in manifests, you can use `kustomize` instead (`-k` option for `kubectl`): https://stackoverflow.com/questions/57594001/how-can-i-create-a-namespace-with-kustomize

Then user needs to update namespace only in two place, the YAML creating the namespace, and in kustomization file.

**Comment by xiguiw** [comment]
@eero-t 

Is this proxy issue of K8S?
What's the status now?

**Comment by eero-t** [comment]
> Is this proxy issue of K8S?

@xiguiw Yes, this bug is k8s specific.

> What's the status now?

K8s manifest files were removed from "GenAIExamples" repo.  It has now only Docker compose, GMC CRD, and Helm values files.

Only things mentioning full k8s service names are GMC README files, and that's not related to proxy.

=> Closing this bug for this repository.

---

"GenAIComps" repo still refers to full service names, which is wrong as their namespace can be wrong:
```
$ git grep svc.cluster.local
comps/cores/mega/exporter.py:            "TEI_EMBEDDING_ENDPOINT": "http://embedding-dependency-svc.default.svc.cluster.local:6006",
comps/cores/mega/exporter.py:            "TEI_ENDPOINT": "http://embedding-dependency-svc.default.svc.cluster.local:6006",
comps/cores/mega/exporter.py:            "REDIS_URL": "redis://vector-db.default.svc.cluster.local:6379",
comps/cores/mega/exporter.py:            "TEI_RERANKING_ENDPOINT": "http://reranking-dependency-svc.default.svc.cluster.local:8808",
comps/cores/mega/exporter.py:            "TGI_LLM_ENDPOINT": "http://llm-dependency-svc.default.svc.cluster.local:9009",
```

And "GenAIInfra" repo has much more of those references, in following subdirs:
```
$ git grep svc.cluster.local | cut -d/ -f1 | uniq -c | sort -nr
     26 microservices-connector
     10 authN-authZ
      1 proxy
      1 kubernetes-addons
```

In different kinds of files:
```
$ git grep svc.cluster.local | cut -d: -f1 | sed 's/^.*\(\.[^.]\+\)$/\1/' | sort | uniq -c | sort -nr
     17 .md
     16 .yaml
      5 .go
```

=> Somebody else can look into those.
